File: frontend/lib/api/jobs.ts
import { api } from "./client";
import type { JobStatusResponse } from "./types";

/**
 * Shared canonical async-job status route — 06-cloud-devops.md §4: "one
 * polling route/hook for every module's async jobs, not a module-specific
 * path." Module 3 uses this for billing-cycle generation beyond ~300 flats
 * (specs/03-maintenance-billing/backend.md §4); future modules reuse it
 * unchanged rather than growing their own polling endpoint.
 */
export function getJobStatus(towerId: string, jobId: string) {
  return api.get<JobStatusResponse>(`/api/v1/towers/${towerId}/jobs/${jobId}`);
}

File: backend/app/services/jobs.py
"""Generic async-job bookkeeping (`app.models.job.Job`) — shared by every module's async paths
per `06-cloud-devops.md` §4. Only Module 3 (billing-cycle generation) populates it today;
Modules 4/5 reuse the same table/route for their own `job_type`s rather than inventing a
module-specific job-status path.
"""

from typing import Any
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.job import Job


async def create_job(
    db: AsyncSession,
    *,
    tower_id: UUID,
    job_type: str,
    payload: dict[str, Any] | None = None,
    idempotency_key: str | None = None,
) -> Job:
    job = Job(
        tower_id=tower_id,
        job_type=job_type,
        status="pending",
        payload=payload,
        idempotency_key=idempotency_key,
    )
    db.add(job)
    await db.flush()
    return job


def mark_running(job: Job) -> None:
    job.status = "running"


def mark_done(job: Job, *, result: dict[str, Any] | None = None) -> None:
    job.status = "done"
    job.result = result


def mark_failed(job: Job, *, error_message: str) -> None:
    job.status = "failed"
    job.error_message = error_message
